// @markuplint/selector/lib/match-selector.js
"use strict";
var _SelectorTarget_combinedFrom, _SelectorTarget_selector;
Object.defineProperty(exports, "__esModule", { value: true });
exports.matchSelector = void 0;
const tslib_1 = require("tslib");
const is_1 = require("./is");
const regex_selector_matches_1 = require("./regex-selector-matches");
const selector_1 = require("./selector");
function matchSelector(
// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
el, selector) {
    if (selector == null || selector === '') {
        return {
            matched: false,
        };
    }
    if (typeof selector === 'string') {
        const sel = new selector_1.Selector(selector);
        const specificity = sel.match(el);
        if (specificity !== false) {
            return {
                matched: true,
                selector,
                specificity,
            };
        }
        return {
            matched: false,
        };
    }
    return regexSelect(el, selector);
}
exports.matchSelector = matchSelector;
function regexSelect(
// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
el, selector) {
    let edge = new SelectorTarget(selector);
    let edgeSelector = selector.combination;
    while (edgeSelector) {
        const child = new SelectorTarget(edgeSelector);
        child.from(edge, edgeSelector.combinator);
        edge = child;
        edgeSelector = edgeSelector.combination;
    }
    return edge.match(el);
}
class SelectorTarget {
    constructor(selector) {
        _SelectorTarget_combinedFrom.set(this, null);
        _SelectorTarget_selector.set(this, void 0);
        tslib_1.__classPrivateFieldSet(this, _SelectorTarget_selector, selector, "f");
    }
    from(target, combinator) {
        tslib_1.__classPrivateFieldSet(this, _SelectorTarget_combinedFrom, { target, combinator }, "f");
    }
    match(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    el) {
        const unitCheck = this._matchWithoutCombineChecking(el);
        if (!unitCheck.matched) {
            return unitCheck;
        }
        if (!tslib_1.__classPrivateFieldGet(this, _SelectorTarget_combinedFrom, "f")) {
            return unitCheck;
        }
        if (!(0, is_1.isNonDocumentTypeChildNode)(el)) {
            return unitCheck;
        }
        const { target, combinator } = tslib_1.__classPrivateFieldGet(this, _SelectorTarget_combinedFrom, "f");
        switch (combinator) {
            // Descendant combinator
            case ' ': {
                let ancestor = el.parentElement;
                while (ancestor) {
                    const matches = target.match(ancestor);
                    if (matches.matched) {
                        return mergeMatches(matches, unitCheck, ' ');
                    }
                    ancestor = ancestor.parentElement;
                }
                return {
                    matched: false,
                };
            }
            // Child combinator
            case '>': {
                const parentNode = el.parentElement;
                if (!parentNode) {
                    return { matched: false };
                }
                const matches = target.match(parentNode);
                if (matches.matched) {
                    return mergeMatches(matches, unitCheck, ' > ');
                }
                return { matched: false };
            }
            // Next-sibling combinator
            case '+': {
                if (!el.previousElementSibling) {
                    return { matched: false };
                }
                const matches = target.match(el.previousElementSibling);
                if (matches.matched) {
                    return mergeMatches(matches, unitCheck, ' + ');
                }
                return { matched: false };
            }
            // Subsequent-sibling combinator
            case '~': {
                let prev = el.previousElementSibling;
                while (prev) {
                    const matches = target.match(prev);
                    if (matches.matched) {
                        return mergeMatches(matches, unitCheck, ' ~ ');
                    }
                    prev = prev.previousElementSibling;
                }
                return { matched: false };
            }
            // Prev-sibling combinator
            case ':has(+)': {
                if (!el.nextElementSibling) {
                    return { matched: false };
                }
                const matches = target.match(el.nextElementSibling);
                if (matches.matched) {
                    return mergeMatches(matches, unitCheck, ':has(+ ', true);
                }
                return { matched: false };
            }
            // Subsequent-sibling (in front) combinator
            case ':has(~)': {
                let next = el.nextElementSibling;
                while (next) {
                    const matches = target.match(next);
                    if (matches.matched) {
                        return mergeMatches(matches, unitCheck, ':has(~ ', true);
                    }
                    next = next.nextElementSibling;
                }
                return { matched: false };
            }
            default: {
                throw new Error(`Unsupported ${tslib_1.__classPrivateFieldGet(this, _SelectorTarget_combinedFrom, "f").combinator} combinator in selector`);
            }
        }
    }
    _matchWithoutCombineChecking(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    el) {
        return uncombinedRegexSelect(el, tslib_1.__classPrivateFieldGet(this, _SelectorTarget_selector, "f"));
    }
}
_SelectorTarget_combinedFrom = new WeakMap(), _SelectorTarget_selector = new WeakMap();
function uncombinedRegexSelect(
// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
el, selector) {
    if (!(0, is_1.isElement)(el)) {
        return {
            matched: false,
        };
    }
    let matched = true;
    let data = {};
    let tagSelector = '';
    const specificity = [0, 0, 0];
    const specifiedAttr = new Map();
    if (selector.nodeName) {
        const matchedNodeName = (0, regex_selector_matches_1.regexSelectorMatches)(selector.nodeName, el.localName, (0, is_1.isPureHTMLElement)(el));
        if (matchedNodeName) {
            delete matchedNodeName.$0;
        }
        else {
            matched = false;
        }
        data = {
            ...data,
            ...matchedNodeName,
        };
        tagSelector = el.localName;
        specificity[2] = 1;
    }
    if (selector.attrName) {
        const selectorAttrName = selector.attrName;
        const matchedAttrNameList = Array.from(el.attributes).map(attr => {
            const attrName = attr.name;
            const matchedAttrName = (0, regex_selector_matches_1.regexSelectorMatches)(selectorAttrName, attrName, (0, is_1.isPureHTMLElement)(el));
            if (matchedAttrName) {
                delete matchedAttrName.$0;
                data = {
                    ...data,
                    ...matchedAttrName,
                };
                specifiedAttr.set(attrName, '');
            }
            return matchedAttrName;
        });
        if (!matchedAttrNameList.some(_ => !!_)) {
            matched = false;
        }
    }
    if (selector.attrValue) {
        const selectorAttrValue = selector.attrValue;
        const matchedAttrValueList = Array.from(el.attributes).map(attr => {
            const attrName = attr.name;
            const attrValue = attr.value;
            const matchedAttrValue = (0, regex_selector_matches_1.regexSelectorMatches)(selectorAttrValue, attrValue, (0, is_1.isPureHTMLElement)(el));
            if (matchedAttrValue) {
                delete matchedAttrValue.$0;
                data = {
                    ...data,
                    ...matchedAttrValue,
                };
                specifiedAttr.set(attrName, attrValue);
            }
            return matchedAttrValue;
        });
        if (!matchedAttrValueList.some(_ => !!_)) {
            matched = false;
        }
    }
    const attrSelector = Array.from(specifiedAttr.entries())
        .map(([name, value]) => {
        return `[${name}${value ? `="${value}"` : ''}]`;
    })
        .join('');
    specificity[1] += specifiedAttr.size;
    if (matched) {
        return {
            matched,
            selector: `${tagSelector}${attrSelector}`,
            specificity,
            data,
        };
    }
    return { matched };
}
function mergeMatches(a, b, sep, close = false) {
    return {
        matched: true,
        selector: `${a.selector}${sep}${b.selector}${close ? ')' : ''}`,
        specificity: [
            a.specificity[0] + b.specificity[0],
            a.specificity[1] + b.specificity[1],
            a.specificity[2] + b.specificity[2],
        ],
        data: {
            ...a.data,
            ...b.data,
        },
    };
}
