// @markuplint/rules/lib/wai-aria/checkings/value.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkAriaValue = exports.checkingValue = void 0;
const checkingValue = ({ attr, role, propSpecs, booleanish }) => t => {
    var _a;
    if (attr.isDynamicValue) {
        return;
    }
    const propSpec = propSpecs.find(p => p.name === attr.name);
    const result = checkAria(propSpec, attr.value, role === null || role === void 0 ? void 0 : role.name, booleanish);
    if (result.isValid) {
        return;
    }
    return {
        scope: attr,
        message: t('{0:c} on {1}', t('{0} is {1:c}', t('the "{0}"', attr.value), 'disallowed'), t('the "{0*}" {1}', attr.name, `ARIA ${(_a = propSpec === null || propSpec === void 0 ? void 0 : propSpec.type) !== null && _a !== void 0 ? _a : 'property'}`)) +
            ('enum' in result && result.enum.length > 0
                ? t('. ') + t('Allowed values are: {0}', t(result.enum))
                : ''),
    };
};
exports.checkingValue = checkingValue;
function checkAria(propSpec, currentValue, role, booleanish) {
    if (!propSpec) {
        return {
            currentValue,
            // For skipping checking
            isValid: true,
        };
    }
    let valueType = propSpec.value;
    if (role && propSpec.conditionalValue) {
        for (const cond of propSpec.conditionalValue) {
            if (cond.role.includes(role)) {
                valueType = cond.value;
                break;
            }
        }
    }
    const isValid = checkAriaValue(valueType, currentValue, propSpec.enum, booleanish);
    return {
        ...propSpec,
        currentValue,
        isValid,
    };
}
/**
 *
 * @see https://www.w3.org/TR/wai-aria-1.2/#propcharacteristic_value
 */
function checkAriaValue(type, value, tokenEnum, booleanish) {
    switch (type) {
        case 'token': {
            return tokenEnum.includes(value);
        }
        case 'token list': {
            const list = value.split(/\s+/g).map(s => s.trim());
            return list.every(token => tokenEnum.includes(token));
        }
        case 'string':
        case 'ID reference':
        case 'ID reference list': {
            return true;
        }
        case 'true/false': {
            if (booleanish && value === '') {
                return true;
            }
            return ['true', 'false'].includes(value);
        }
        case 'tristate': {
            if (booleanish && value === '') {
                return true;
            }
            return ['mixed', 'true', 'false', 'undefined'].includes(value);
        }
        case 'true/false/undefined': {
            if (booleanish && value === '') {
                return true;
            }
            return ['true', 'false', 'undefined'].includes(value);
        }
        case 'integer': {
            return parseInt(value).toString() === value;
        }
        case 'number': {
            return parseFloat(value).toString() === value;
        }
    }
    // For skipping checking
    return true;
}
exports.checkAriaValue = checkAriaValue;
