// @markuplint/rules/lib/required-element/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ml_core_1 = require("@markuplint/ml-core");
exports.default = (0, ml_core_1.createRule)({
    defaultValue: [],
    defaultOptions: {
        ignoreHasMutableContents: true,
    },
    async verify({ document, report, t }) {
        var _a, _b;
        const hasMutableContent = document.nodeList.some(n => n.is(n.ELEMENT_NODE) && n.hasMutableChildren());
        if (!hasMutableContent) {
            for (const query of document.rule.value) {
                const exists = document.querySelectorAll(query);
                if (exists.length === 0) {
                    const message = t('Require {0}', t('the "{0*}" {1}', query, 'element'));
                    report({
                        message,
                        line: 1,
                        col: 1,
                        raw: (_b = (_a = document.nodeList[0]) === null || _a === void 0 ? void 0 : _a.raw.slice(0, 1)) !== null && _b !== void 0 ? _b : '',
                    });
                }
            }
        }
        await document.walkOn('Element', el => {
            if (el.rule.value === document.rule.value) {
                return;
            }
            if (el.rule.options.ignoreHasMutableContents && el.hasMutableChildren()) {
                return;
            }
            for (const query of el.rule.value) {
                const exists = Array.from(el.children).find(child => child.matches(query));
                if (!exists) {
                    const message = t('Require {0}', t('the "{0*}" {1}', query, 'element'));
                    report({
                        scope: el,
                        message,
                    });
                }
            }
        });
    },
});
