// @markuplint/ml-spec/lib/utils/resolve-namespace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveNamespace = void 0;
const get_ns_1 = require("./get-ns");
const cache = new Map();
const namespaceURIMap = {
    html: 'http://www.w3.org/1999/xhtml',
    svg: 'http://www.w3.org/2000/svg',
    mml: 'http://www.w3.org/1998/Math/MathML',
    xlink: 'http://www.w3.org/1999/xlink',
};
function resolveNamespace(name, namespaceURI = 'http://www.w3.org/1999/xhtml') {
    var _a;
    const cached = cache.get(name + namespaceURI);
    if (cached) {
        return cached;
    }
    const [_explicitNS, _localName] = name.split(':');
    const explicitNS = _localName ? _explicitNS : null;
    const localName = (_a = _localName !== null && _localName !== void 0 ? _localName : _explicitNS) !== null && _a !== void 0 ? _a : '';
    const namespace = ['html', 'svg', 'mml', 'xlink'].find(_ns => _ns === (explicitNS || (0, get_ns_1.getNS)(namespaceURI !== null && namespaceURI !== void 0 ? namespaceURI : null))) ||
        'html';
    const result = {
        localNameWithNS: `${namespace === 'html' ? '' : `${namespace}:`}${localName}`,
        localName,
        namespace,
        namespaceURI: namespaceURIMap[namespace],
    };
    cache.set(name + namespaceURI, result);
    return result;
}
exports.resolveNamespace = resolveNamespace;
