// @markuplint/ml-core/lib/ml-dom/node/element.js
"use strict";
/* global StylePropertyMap, StylePropertyMapReadOnly */
var _MLElement_attributes, _MLElement_fixedNodeName, _MLElement_getChildElementsAndTextNodeWithoutWhitespacesCache, _MLElement_localName, _MLElement_normalizedAttrs, _MLElement_normalizedString, _MLElement_tagOpenChar;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MLElement = void 0;
const tslib_1 = require("tslib");
const ml_spec_1 = require("@markuplint/ml-spec");
const selector_1 = require("@markuplint/selector");
const string_splice_1 = require("../../utils/string-splice");
const accname_1 = require("../helper/accname");
const child_node_methods_1 = require("../manipulations/child-node-methods");
const token_1 = require("../token/token");
const attr_1 = require("./attr");
const dom_token_list_1 = require("./dom-token-list");
const named_node_map_1 = require("./named-node-map");
const node_list_1 = require("./node-list");
const parent_node_1 = require("./parent-node");
const unexpected_call_error_1 = tslib_1.__importDefault(require("./unexpected-call-error"));
const HTML_NAMESPACE = 'http://www.w3.org/1999/xhtml';
class MLElement extends parent_node_1.MLParentNode {
    constructor(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    astNode, 
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    document) {
        super(astNode, document);
        _MLElement_attributes.set(this, void 0);
        _MLElement_fixedNodeName.set(this, void 0);
        _MLElement_getChildElementsAndTextNodeWithoutWhitespacesCache.set(this, null);
        _MLElement_localName.set(this, void 0);
        _MLElement_normalizedAttrs.set(this, new Map());
        _MLElement_normalizedString.set(this, null);
        this.pretenderContext = null;
        _MLElement_tagOpenChar.set(this, void 0);
        tslib_1.__classPrivateFieldSet(this, _MLElement_attributes, astNode.attributes.map(attr => new attr_1.MLAttr(attr, this)), "f");
        this.hasSpreadAttr = astNode.hasSpreadAttr;
        this.selfClosingSolidus = astNode.selfClosingSolidus ? new token_1.MLToken(astNode.selfClosingSolidus) : null;
        this.endSpace = astNode.endSpace ? new token_1.MLToken(astNode.endSpace) : null;
        this.closeTag = astNode.pearNode ? new token_1.MLToken(astNode.pearNode) : null;
        const ns = (0, ml_spec_1.resolveNamespace)(astNode.nodeName, astNode.namespace);
        this.namespaceURI = ns.namespaceURI;
        this.elementType = astNode.elementType;
        tslib_1.__classPrivateFieldSet(this, _MLElement_localName, ns.localName, "f");
        this.isForeignElement = this.namespaceURI !== HTML_NAMESPACE;
        tslib_1.__classPrivateFieldSet(this, _MLElement_fixedNodeName, astNode.nodeName, "f");
        this.isOmitted = astNode.isGhost;
        tslib_1.__classPrivateFieldSet(this, _MLElement_tagOpenChar, astNode.tagOpenChar, "f");
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get accessKey() {
        throw new unexpected_call_error_1.default('Not supported "accessKey" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get accessKeyLabel() {
        throw new unexpected_call_error_1.default('Not supported "accessKeyLabel" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaAtomic() {
        throw new unexpected_call_error_1.default('Not supported "ariaAtomic" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaAutoComplete() {
        throw new unexpected_call_error_1.default('Not supported "ariaAutoComplete" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaBusy() {
        throw new unexpected_call_error_1.default('Not supported "ariaBusy" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaChecked() {
        throw new unexpected_call_error_1.default('Not supported "ariaChecked" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaColCount() {
        throw new unexpected_call_error_1.default('Not supported "ariaColCount" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaColIndex() {
        throw new unexpected_call_error_1.default('Not supported "ariaColIndex" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaColIndexText() {
        throw new unexpected_call_error_1.default('Not supported "ariaColIndexText" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaColSpan() {
        throw new unexpected_call_error_1.default('Not supported "ariaColSpan" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaCurrent() {
        throw new unexpected_call_error_1.default('Not supported "ariaCurrent" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaDisabled() {
        throw new unexpected_call_error_1.default('Not supported "ariaDisabled" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaExpanded() {
        throw new unexpected_call_error_1.default('Not supported "ariaExpanded" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaHasPopup() {
        throw new unexpected_call_error_1.default('Not supported "ariaHasPopup" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaHidden() {
        throw new unexpected_call_error_1.default('Not supported "ariaHidden" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaInvalid() {
        throw new unexpected_call_error_1.default('Not supported "ariaInvalid" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaKeyShortcuts() {
        throw new unexpected_call_error_1.default('Not supported "ariaKeyShortcuts" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaLabel() {
        throw new unexpected_call_error_1.default('Not supported "ariaLabel" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaLevel() {
        throw new unexpected_call_error_1.default('Not supported "ariaLevel" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaLive() {
        throw new unexpected_call_error_1.default('Not supported "ariaLive" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaModal() {
        throw new unexpected_call_error_1.default('Not supported "ariaModal" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaMultiLine() {
        throw new unexpected_call_error_1.default('Not supported "ariaMultiLine" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaMultiSelectable() {
        throw new unexpected_call_error_1.default('Not supported "ariaMultiSelectable" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaOrientation() {
        throw new unexpected_call_error_1.default('Not supported "ariaOrientation" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaPlaceholder() {
        throw new unexpected_call_error_1.default('Not supported "ariaPlaceholder" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaPosInSet() {
        throw new unexpected_call_error_1.default('Not supported "ariaPosInSet" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaPressed() {
        throw new unexpected_call_error_1.default('Not supported "ariaPressed" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaReadOnly() {
        throw new unexpected_call_error_1.default('Not supported "ariaReadOnly" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaRequired() {
        throw new unexpected_call_error_1.default('Not supported "ariaRequired" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaRoleDescription() {
        throw new unexpected_call_error_1.default('Not supported "ariaRoleDescription" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaRowCount() {
        throw new unexpected_call_error_1.default('Not supported "ariaRowCount" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaRowIndex() {
        throw new unexpected_call_error_1.default('Not supported "ariaRowIndex" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaRowIndexText() {
        throw new unexpected_call_error_1.default('Not supported "ariaRowIndexText" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaRowSpan() {
        throw new unexpected_call_error_1.default('Not supported "ariaRowSpan" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaSelected() {
        throw new unexpected_call_error_1.default('Not supported "ariaSelected" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaSetSize() {
        throw new unexpected_call_error_1.default('Not supported "ariaSetSize" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaSort() {
        throw new unexpected_call_error_1.default('Not supported "ariaSort" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaValueMax() {
        throw new unexpected_call_error_1.default('Not supported "ariaValueMax" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaValueMin() {
        throw new unexpected_call_error_1.default('Not supported "ariaValueMin" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaValueNow() {
        throw new unexpected_call_error_1.default('Not supported "ariaValueNow" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ariaValueText() {
        throw new unexpected_call_error_1.default('Not supported "ariaValueText" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#dom-slotable-assignedslot
     */
    get assignedSlot() {
        throw new unexpected_call_error_1.default('Not supported "assignedSlot" property');
    }
    get attributeStyleMap() {
        throw new unexpected_call_error_1.default('Not supported "attributeStyleMap" property');
    }
    /**
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#dom-element-attributes
     */
    get attributes() {
        var _a;
        const origin = ((_a = this.pretenderContext) === null || _a === void 0 ? void 0 : _a.type) === 'pretender' ? tslib_1.__classPrivateFieldGet(this.pretenderContext.as, _MLElement_attributes, "f") : tslib_1.__classPrivateFieldGet(this, _MLElement_attributes, "f");
        if (tslib_1.__classPrivateFieldGet(this, _MLElement_normalizedAttrs, "f").has(origin)) {
            return tslib_1.__classPrivateFieldGet(this, _MLElement_normalizedAttrs, "f").get(origin);
        }
        const names = new Set();
        const attrs = [];
        for (const attr of origin) {
            if (names.has(attr.name)) {
                /**
                 * Skips a duplicated attribute
                 *
                 *@see https://html.spec.whatwg.org/#parse-error-duplicate-attribute
                 */
                continue;
            }
            attrs.push(attr);
        }
        const map = (0, named_node_map_1.toNamedNodeMap)(attrs);
        tslib_1.__classPrivateFieldGet(this, _MLElement_normalizedAttrs, "f").set(origin, map);
        return map;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get autocapitalize() {
        throw new unexpected_call_error_1.default('Not supported "autocapitalize" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get autofocus() {
        throw new unexpected_call_error_1.default('Not supported "autofocus" property');
    }
    /**
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-classlist%E2%91%A0
     */
    get classList() {
        const classAttrs = this.getAttributeToken('class');
        const value = classAttrs.map(c => c.value).join(' ');
        return new dom_token_list_1.MLDomTokenList(value, classAttrs);
    }
    /**
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-classname%E2%91%A0
     */
    get className() {
        return this.classList.value;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/cssom-view-1/#ref-for-dom-element-clientheight-1
     */
    get clientHeight() {
        throw new unexpected_call_error_1.default('Not supported "clientHeight" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/cssom-view-1/#ref-for-dom-element-clientleft-1
     */
    get clientLeft() {
        throw new unexpected_call_error_1.default('Not supported "clientLeft" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/cssom-view-1/#ref-for-dom-element-clienttop-1
     */
    get clientTop() {
        throw new unexpected_call_error_1.default('Not supported "clientTop" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/cssom-view-1/#ref-for-dom-element-clientwidth-1
     */
    get clientWidth() {
        throw new unexpected_call_error_1.default('Not supported "clientWidth" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get contentEditable() {
        throw new unexpected_call_error_1.default('Not supported "contentEditable" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get dataset() {
        throw new unexpected_call_error_1.default('Not supported "dataset" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get dir() {
        throw new unexpected_call_error_1.default('Not supported "dir" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get draggable() {
        throw new unexpected_call_error_1.default('Not supported "draggable" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get enterKeyHint() {
        throw new unexpected_call_error_1.default('Not supported "enterKeyHint" property');
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLElement`
     */
    get fixedNodeName() {
        return tslib_1.__classPrivateFieldGet(this, _MLElement_fixedNodeName, "f");
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get hidden() {
        throw new unexpected_call_error_1.default('Not supported "hidden" property');
    }
    /**
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-id%E2%91%A0
     */
    get id() {
        var _a;
        return (_a = this.getAttribute('id')) !== null && _a !== void 0 ? _a : '';
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://html.spec.whatwg.org/multipage/interaction.html#dom-inert
     */
    get inert() {
        throw new unexpected_call_error_1.default('Not supported "inert" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://w3c.github.io/DOM-Parsing/#dom-innerhtml-innerhtml
     */
    get innerHTML() {
        throw new unexpected_call_error_1.default('Not supported "innerHTML" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get innerText() {
        throw new unexpected_call_error_1.default('Not supported "innerText" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get inputMode() {
        throw new unexpected_call_error_1.default('Not supported "inputMode" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get isContentEditable() {
        throw new unexpected_call_error_1.default('Not supported "isContentEditable" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get lang() {
        throw new unexpected_call_error_1.default('Not supported "lang" property');
    }
    /**
     * Returns a lowercase name if it is an HTML element.
     *
     * ```html
     * <div> => "div"
     * <DIV> => "div"
     * <svg> => "svg"
     * <foreignObject> => "foreignObject"
     * <x-foo> => "x-foo"
     * <X-FOO> => "x-foo"
     * ```
     *
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-localname%E2%91%A0
     */
    get localName() {
        var _a;
        if (((_a = this.pretenderContext) === null || _a === void 0 ? void 0 : _a.type) === 'pretender') {
            return this.pretenderContext.as.localName;
        }
        if (this.isForeignElement || this.elementType !== 'html') {
            return tslib_1.__classPrivateFieldGet(this, _MLElement_localName, "f");
        }
        return tslib_1.__classPrivateFieldGet(this, _MLElement_localName, "f").toLowerCase();
    }
    /**
     * The element immediately following the specified one in its parent's children list.
     *
     * @readonly
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-nondocumenttypechildnode-nextelementsibling%E2%91%A1
     */
    get nextElementSibling() {
        return (0, child_node_methods_1.nextElementSibling)(this);
    }
    /**
     * Returns a string appropriate for the type of node as `Element`
     *
     * @see https://dom.spec.whatwg.org/#ref-for-element%E2%91%A2%E2%93%AA
     */
    get nodeName() {
        var _a;
        if (((_a = this.pretenderContext) === null || _a === void 0 ? void 0 : _a.type) === 'pretender') {
            return this.pretenderContext.as.nodeName;
        }
        if (this.isForeignElement || this.elementType !== 'html') {
            return this._astToken.nodeName;
        }
        return this._astToken.nodeName.toUpperCase();
    }
    /**
     * Returns a number appropriate for the type of `Element`
     */
    get nodeType() {
        return this.ELEMENT_NODE;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get nonce() {
        throw new unexpected_call_error_1.default('Not supported "nonce" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get offsetHeight() {
        throw new unexpected_call_error_1.default('Not supported "offsetHeight" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get offsetLeft() {
        throw new unexpected_call_error_1.default('Not supported "offsetLeft" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get offsetParent() {
        throw new unexpected_call_error_1.default('Not supported "offsetParent" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get offsetTop() {
        throw new unexpected_call_error_1.default('Not supported "offsetTop" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get offsetWidth() {
        throw new unexpected_call_error_1.default('Not supported "offsetWidth" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onabort() {
        throw new unexpected_call_error_1.default('Not supported "onabort" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onanimationcancel() {
        throw new unexpected_call_error_1.default('Not supported "onanimationcancel" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onanimationend() {
        throw new unexpected_call_error_1.default('Not supported "onanimationend" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onanimationiteration() {
        throw new unexpected_call_error_1.default('Not supported "onanimationiteration" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onanimationstart() {
        throw new unexpected_call_error_1.default('Not supported "onanimationstart" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onauxclick() {
        throw new unexpected_call_error_1.default('Not supported "onauxclick" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onbeforeinput() {
        throw new unexpected_call_error_1.default('Not supported "onbeforeinput" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onblur() {
        throw new unexpected_call_error_1.default('Not supported "onblur" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get oncancel() {
        throw new unexpected_call_error_1.default('Not supported "oncanplay" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get oncanplay() {
        throw new unexpected_call_error_1.default('Not supported "oncanplay" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get oncanplaythrough() {
        throw new unexpected_call_error_1.default('Not supported "oncanplaythrough" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onchange() {
        throw new unexpected_call_error_1.default('Not supported "onchange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onclick() {
        throw new unexpected_call_error_1.default('Not supported "onclick" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onclose() {
        throw new unexpected_call_error_1.default('Not supported "onclose" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get oncontextmenu() {
        throw new unexpected_call_error_1.default('Not supported "oncontextmenu" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get oncopy() {
        throw new unexpected_call_error_1.default('Not supported "oncopy" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get oncuechange() {
        throw new unexpected_call_error_1.default('Not supported "onpointerleave" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get oncut() {
        throw new unexpected_call_error_1.default('Not supported "onpointerleave" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ondblclick() {
        throw new unexpected_call_error_1.default('Not supported "ondblclick" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ondrag() {
        throw new unexpected_call_error_1.default('Not supported "ondrag" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ondragend() {
        throw new unexpected_call_error_1.default('Not supported "ondragend" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ondragenter() {
        throw new unexpected_call_error_1.default('Not supported "ondragenter" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ondragleave() {
        throw new unexpected_call_error_1.default('Not supported "ondragleave" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ondragover() {
        throw new unexpected_call_error_1.default('Not supported "ondragover" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ondragstart() {
        throw new unexpected_call_error_1.default('Not supported "ondragstart" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ondrop() {
        throw new unexpected_call_error_1.default('Not supported "ondrop" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ondurationchange() {
        throw new unexpected_call_error_1.default('Not supported "ondurationchange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onemptied() {
        throw new unexpected_call_error_1.default('Not supported "onemptied" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onended() {
        throw new unexpected_call_error_1.default('Not supported "onended" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onerror() {
        throw new unexpected_call_error_1.default('Not supported "onerror" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onfocus() {
        throw new unexpected_call_error_1.default('Not supported "onfocus" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onformdata() {
        throw new unexpected_call_error_1.default('Not supported "onformdata" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onfullscreenchange() {
        throw new unexpected_call_error_1.default('Not supported "onfullscreenchange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onfullscreenerror() {
        throw new unexpected_call_error_1.default('Not supported "onfullscreenerror" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ongotpointercapture() {
        throw new unexpected_call_error_1.default('Not supported "ongotpointercapture" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get oninput() {
        throw new unexpected_call_error_1.default('Not supported "oninput" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get oninvalid() {
        throw new unexpected_call_error_1.default('Not supported "oninvalid" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onkeydown() {
        throw new unexpected_call_error_1.default('Not supported "onkeydown" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onkeypress() {
        throw new unexpected_call_error_1.default('Not supported "onkeypress" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onkeyup() {
        throw new unexpected_call_error_1.default('Not supported "onkeyup" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onload() {
        throw new unexpected_call_error_1.default('Not supported "onload" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onloadeddata() {
        throw new unexpected_call_error_1.default('Not supported "onloadeddata" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onloadedmetadata() {
        throw new unexpected_call_error_1.default('Not supported "onloadedmetadata" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onloadstart() {
        throw new unexpected_call_error_1.default('Not supported "onloadstart" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onlostpointercapture() {
        throw new unexpected_call_error_1.default('Not supported "onlostpointercapture" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onmousedown() {
        throw new unexpected_call_error_1.default('Not supported "onmousedown" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onmouseenter() {
        throw new unexpected_call_error_1.default('Not supported "onmouseenter" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onmouseleave() {
        throw new unexpected_call_error_1.default('Not supported "onmouseleave" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onmousemove() {
        throw new unexpected_call_error_1.default('Not supported "onmousemove" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onmouseout() {
        throw new unexpected_call_error_1.default('Not supported "onmouseout" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onmouseover() {
        throw new unexpected_call_error_1.default('Not supported "onmouseover" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onmouseup() {
        throw new unexpected_call_error_1.default('Not supported "onmouseup" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onpaste() {
        throw new unexpected_call_error_1.default('Not supported "onpaste" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onpause() {
        throw new unexpected_call_error_1.default('Not supported "onpause" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onplay() {
        throw new unexpected_call_error_1.default('Not supported "onplay" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onplaying() {
        throw new unexpected_call_error_1.default('Not supported "onplaying" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onpointercancel() {
        throw new unexpected_call_error_1.default('Not supported "onpointercancel" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onpointerdown() {
        throw new unexpected_call_error_1.default('Not supported "onpointerdown" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onpointerenter() {
        throw new unexpected_call_error_1.default('Not supported "onpointerenter" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onpointerleave() {
        throw new unexpected_call_error_1.default('Not supported "onpointerleave" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onpointermove() {
        throw new unexpected_call_error_1.default('Not supported "onpointermove" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onpointerout() {
        throw new unexpected_call_error_1.default('Not supported "onpointerout" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onpointerover() {
        throw new unexpected_call_error_1.default('Not supported "onpointerover" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onpointerup() {
        throw new unexpected_call_error_1.default('Not supported "onpointerup" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onprogress() {
        throw new unexpected_call_error_1.default('Not supported "onprogress" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onratechange() {
        throw new unexpected_call_error_1.default('Not supported "onratechange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onreset() {
        throw new unexpected_call_error_1.default('Not supported "onreset" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onresize() {
        throw new unexpected_call_error_1.default('Not supported "onresize" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onscroll() {
        throw new unexpected_call_error_1.default('Not supported "onscroll" property');
    }
    get onscrollend() {
        throw new unexpected_call_error_1.default('Not supported "onscrollend" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onsecuritypolicyviolation() {
        throw new unexpected_call_error_1.default('Not supported "onsecuritypolicyviolation" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onseeked() {
        throw new unexpected_call_error_1.default('Not supported "onseeked" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onseeking() {
        throw new unexpected_call_error_1.default('Not supported "onseeking" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onselect() {
        throw new unexpected_call_error_1.default('Not supported "onselect" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onselectionchange() {
        throw new unexpected_call_error_1.default('Not supported "onselectionchange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onselectstart() {
        throw new unexpected_call_error_1.default('Not supported "onselectstart" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onslotchange() {
        throw new unexpected_call_error_1.default('Not supported "onslotchange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onstalled() {
        throw new unexpected_call_error_1.default('Not supported "onstalled" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onsubmit() {
        throw new unexpected_call_error_1.default('Not supported "onsubmit" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onsuspend() {
        throw new unexpected_call_error_1.default('Not supported "onsuspend" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ontimeupdate() {
        throw new unexpected_call_error_1.default('Not supported "ontimeupdate" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ontoggle() {
        throw new unexpected_call_error_1.default('Not supported "ontoggle" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ontransitioncancel() {
        throw new unexpected_call_error_1.default('Not supported "ontransitioncancel" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ontransitionend() {
        throw new unexpected_call_error_1.default('Not supported "ontransitionend" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ontransitionrun() {
        throw new unexpected_call_error_1.default('Not supported "ontransitionrun" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get ontransitionstart() {
        throw new unexpected_call_error_1.default('Not supported "ontransitionstart" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onvolumechange() {
        throw new unexpected_call_error_1.default('Not supported "onvolumechange" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onwaiting() {
        throw new unexpected_call_error_1.default('Not supported "onwaiting" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onwebkitanimationend() {
        throw new unexpected_call_error_1.default('Not supported "onwebkitanimationend" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onwebkitanimationiteration() {
        throw new unexpected_call_error_1.default('Not supported "onwebkitanimationiteration" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onwebkitanimationstart() {
        throw new unexpected_call_error_1.default('Not supported "onwebkitanimationstart" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onwebkittransitionend() {
        throw new unexpected_call_error_1.default('Not supported "onwebkittransitionend" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get onwheel() {
        throw new unexpected_call_error_1.default('Not supported "onwheel" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://w3c.github.io/DOM-Parsing/#dom-element-outerhtml
     */
    get outerHTML() {
        throw new unexpected_call_error_1.default('Not supported "outerHTML" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get outerText() {
        throw new unexpected_call_error_1.default('Not supported "outerText" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/css-shadow-parts-1/#idl
     */
    get part() {
        throw new unexpected_call_error_1.default('Not supported "part" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://html.spec.whatwg.org/multipage/popover.html#dom-popover
     */
    get popover() {
        throw new unexpected_call_error_1.default('Not supported "popover" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-prefix%E2%91%A0
     */
    get prefix() {
        throw new unexpected_call_error_1.default('Not supported "prefix" property');
    }
    /**
     * The element immediately prior the specified one in its parent's children list.
     *
     * @readonly
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-nondocumenttypechildnode-previouselementsibling%E2%91%A1
     */
    get previousElementSibling() {
        return (0, child_node_methods_1.previousElementSibling)(this);
    }
    get raw() {
        var _a, _b, _c;
        if (((_a = this.pretenderContext) === null || _a === void 0 ? void 0 : _a.type) === 'pretender') {
            return this.originRaw;
        }
        let fixed = this.originRaw;
        let gap = 0;
        if (this.nodeName !== this.fixedNodeName) {
            fixed = (0, string_splice_1.stringSplice)(fixed, tslib_1.__classPrivateFieldGet(this, _MLElement_tagOpenChar, "f").length, this.nodeName.length, this.fixedNodeName);
            gap = gap + this.fixedNodeName.length - this.nodeName.length;
        }
        for (const attr of Array.from(this.attributes)) {
            const startOffset = ((_c = (_b = attr.spacesBeforeName) === null || _b === void 0 ? void 0 : _b.startOffset) !== null && _c !== void 0 ? _c : attr.startOffset) - this.startOffset;
            const fixedAttr = attr.toString();
            if (attr.originRaw !== fixedAttr) {
                fixed = (0, string_splice_1.stringSplice)(fixed, startOffset + gap, attr.originRaw.length, fixedAttr);
                gap = gap + fixedAttr.length - attr.originRaw.length;
            }
        }
        return fixed;
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLElement`
     */
    get rawName() {
        return this._astToken.nodeName;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://w3c.github.io/aria/#ARIAMixin
     */
    get role() {
        throw new unexpected_call_error_1.default('Not supported "role" property');
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLNode`
     */
    get rule() {
        var _a;
        if (((_a = this.pretenderContext) === null || _a === void 0 ? void 0 : _a.type) === 'origin') {
            return this.pretenderContext.origin.rule;
        }
        return super.rule;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/cssom-view-1/#ref-for-dom-element-scrollheight-1
     */
    get scrollHeight() {
        throw new unexpected_call_error_1.default('Not supported "scrollHeight" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/cssom-view-1/#ref-for-dom-element-scrollleft-1
     */
    get scrollLeft() {
        throw new unexpected_call_error_1.default('Not supported "scrollLeft" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/cssom-view-1/#ref-for-dom-element-scrolltop-1
     */
    get scrollTop() {
        throw new unexpected_call_error_1.default('Not supported "scrollTop" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/cssom-view-1/#ref-for-dom-element-scrollwidth-1
     */
    get scrollWidth() {
        throw new unexpected_call_error_1.default('Not supported "scrollWidth" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-shadowroot%E2%91%A0
     */
    get shadowRoot() {
        throw new unexpected_call_error_1.default('Not supported "shadowRoot" property');
    }
    /**
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-slot%E2%91%A0
     */
    get slot() {
        var _a;
        return (_a = this.getAttribute('slot')) !== null && _a !== void 0 ? _a : '';
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get spellcheck() {
        throw new unexpected_call_error_1.default('Not supported "spellcheck" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get style() {
        throw new unexpected_call_error_1.default('Not supported "style" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get tabIndex() {
        throw new unexpected_call_error_1.default('Not supported "tabIndex" property');
    }
    /**
     * Returns the HTML-uppercased qualified name.
     *
     * If this is in the HTML namespace and its node document is an HTML document, then set qualifiedName to qualifiedName in ASCII uppercase.
     *
     * (In markuplint evaluation, the node document always is an HTML document.)
     *
     * ```html
     * <div> => "DIV"
     * <svg> => "svg"
     * <x-foo> => "X-FOO"
     * ```
     *
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-tagname%E2%91%A0
     */
    get tagName() {
        var _a;
        if (((_a = this.pretenderContext) === null || _a === void 0 ? void 0 : _a.type) === 'pretender') {
            return this.pretenderContext.as.nodeName;
        }
        return this.nodeName;
    }
    get textContent() {
        return Array.from(this.childNodes)
            .map(child => { var _a; return (_a = child.textContent) !== null && _a !== void 0 ? _a : ''; })
            .join('');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get title() {
        throw new unexpected_call_error_1.default('Not supported "title" property');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    get translate() {
        throw new unexpected_call_error_1.default('Not supported "translate" property');
    }
    /**
     * @implements DOM API: `Element`
     */
    after(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    ...nodes) {
        (0, child_node_methods_1.after)(this, ...nodes);
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/web-animations-1/#dom-animatable-animate
     */
    animate(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    keyframes, 
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    options) {
        throw new unexpected_call_error_1.default('Not supported "animate" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    attachInternals() {
        throw new unexpected_call_error_1.default('Not supported "attachInternals" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-attachshadow%E2%91%A0
     */
    attachShadow(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    init) {
        throw new unexpected_call_error_1.default('Not supported "attachShadow" method');
    }
    /**
     * @implements DOM API: `Element`
     */
    before(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    ...nodes) {
        (0, child_node_methods_1.before)(this, ...nodes);
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    blur() {
        throw new unexpected_call_error_1.default('Not supported "blur" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    checkVisibility(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    options) {
        throw new unexpected_call_error_1.default('Not supported "checkVisibility" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    click() {
        throw new unexpected_call_error_1.default('Not supported "click" method');
    }
    /**
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-closest%E2%91%A0
     */
    closest(selectors) {
        let el = this;
        do {
            if (el.matches(selectors)) {
                return el;
            }
            el = el.parentElement;
        } while (el !== null && el.is(el.ELEMENT_NODE));
        return null;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     */
    computedStyleMap() {
        throw new unexpected_call_error_1.default('Not supported "computedStyleMap" method');
    }
    fixNodeName(name) {
        tslib_1.__classPrivateFieldSet(this, _MLElement_fixedNodeName, name, "f");
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    focus(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    options) {
        throw new unexpected_call_error_1.default('Not supported "focus" method');
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLElement`
     */
    getAccessibleName(version) {
        return (0, accname_1.getAccname)(this, version);
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/web-animations-1/#dom-animatable-getanimations
     */
    getAnimations(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    options) {
        throw new unexpected_call_error_1.default('Not supported "getAnimations" method');
    }
    /**
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-getattribute%E2%91%A0
     */
    getAttribute(attrName) {
        for (const attr of Array.from(this.attributes)) {
            if (attr.name.toLowerCase() === attrName.toLowerCase()) {
                return attr.value;
            }
        }
        return null;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#dom-element-getattributenodens
     */
    getAttributeNS(namespace, localName) {
        throw new unexpected_call_error_1.default('Not supported "getAttributeNS" method');
    }
    /**
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-getattributenames%E2%91%A0
     */
    getAttributeNames() {
        return Array.from(this.attributes).map(attr => attr.name);
    }
    /**
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#dom-element-getattributenode
     */
    getAttributeNode(qualifiedName) {
        var _a;
        return (_a = this.getAttributeToken(qualifiedName)[0]) !== null && _a !== void 0 ? _a : null;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#dom-element-getattributenodens
     */
    getAttributeNodeNS(namespace, localName) {
        throw new unexpected_call_error_1.default('Not supported "getAttributeNodeNS" method');
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLElement`
     */
    getAttributePretended(attrName) {
        for (const attr of Array.from(tslib_1.__classPrivateFieldGet(this, _MLElement_attributes, "f"))) {
            if (attr.name.toLowerCase() === attrName.toLowerCase()) {
                return attr.value;
            }
        }
        return null;
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLElement`
     */
    getAttributeToken(attrName) {
        const attrs = [];
        attrName = attrName.toLowerCase();
        for (const attr of this.getAttributeTokens()) {
            if (attr.name === attrName) {
                attrs.push(attr);
            }
        }
        return attrs;
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLElement`
     */
    getAttributeTokens() {
        var _a;
        return Object.freeze(((_a = this.pretenderContext) === null || _a === void 0 ? void 0 : _a.type) === 'pretender' ? tslib_1.__classPrivateFieldGet(this.pretenderContext.as, _MLElement_attributes, "f") : tslib_1.__classPrivateFieldGet(this, _MLElement_attributes, "f"));
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/cssom-view-1/#ref-for-dom-element-getboundingclientrect-1
     */
    getBoundingClientRect() {
        throw new unexpected_call_error_1.default('Not supported "getBoundingClientRect" method');
    }
    getChildElementsAndTextNodeWithoutWhitespaces() {
        if (tslib_1.__classPrivateFieldGet(this, _MLElement_getChildElementsAndTextNodeWithoutWhitespacesCache, "f")) {
            return tslib_1.__classPrivateFieldGet(this, _MLElement_getChildElementsAndTextNodeWithoutWhitespacesCache, "f");
        }
        const filteredNodes = [];
        this.childNodes.forEach(node => {
            if (node.is(node.ELEMENT_NODE)) {
                if (node.isOmitted) {
                    const children = node.getChildElementsAndTextNodeWithoutWhitespaces();
                    filteredNodes.push(...children);
                }
                else {
                    filteredNodes.push(node);
                }
            }
            if (node.is(node.TEXT_NODE) && !node.isWhitespace()) {
                filteredNodes.push(node);
            }
        });
        tslib_1.__classPrivateFieldSet(this, _MLElement_getChildElementsAndTextNodeWithoutWhitespacesCache, filteredNodes, "f");
        return filteredNodes;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/cssom-view-1/#ref-for-dom-element-getclientrects-1
     */
    getClientRects() {
        throw new unexpected_call_error_1.default('Not supported "getClientRects" method');
    }
    /**
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-getelementsbyclassname
     */
    getElementsByClassName(classNames) {
        return (0, node_list_1.toHTMLCollection)(this._descendantsToArray(node => {
            if (node.is(node.ELEMENT_NODE) && node.classList.contains(classNames)) {
                return node;
            }
        }));
    }
    /**
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-getelementsbyclassname
     */
    getElementsByTagName(qualifiedName) {
        return (0, node_list_1.toHTMLCollection)(this._descendantsToArray(node => {
            if (node.is(node.ELEMENT_NODE) && node.nodeName.toLowerCase() === qualifiedName.toLowerCase()) {
                return node;
            }
        }));
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#dom-element-getelementsbytagnamens
     */
    getElementsByTagNameNS(namespace, localName) {
        throw new unexpected_call_error_1.default('Not supported "getElementsByTagNameNS" method');
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLElement`
     */
    getNameLocation() {
        return {
            offset: this.startOffset,
            line: this.startLine,
            col: this.startCol + tslib_1.__classPrivateFieldGet(this, _MLElement_tagOpenChar, "f").length,
        };
    }
    /**
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-hasattribute%E2%91%A0
     */
    hasAttribute(qualifiedName) {
        return this.getAttribute(qualifiedName) !== null;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-hasattributens%E2%91%A0
     */
    hasAttributeNS(namespace, localName) {
        throw new unexpected_call_error_1.default('Not supported "hasAttributeNS" method');
    }
    /**
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#dom-element-hasattributes
     */
    hasAttributes() {
        return this.attributes.length > 0;
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLElement`
     */
    hasMutableAttributes() {
        for (const attr of Array.from(this.attributes)) {
            if (!attr.nameNode) {
                return true;
            }
            if (attr.isDynamicValue) {
                return true;
            }
        }
        return false;
    }
    /**
     * This element has "Preprocessor Specific Block". In other words, Its children are potentially mutable.
     *
     * @implements `@markuplint/ml-core` API: `MLElement`
     */
    hasMutableChildren(attr = false) {
        for (const child of Array.from(this.childNodes)) {
            if (child.is(child.MARKUPLINT_PREPROCESSOR_BLOCK)) {
                return true;
            }
            if (child.is(child.ELEMENT_NODE)) {
                if (attr && child.hasMutableAttributes()) {
                    return true;
                }
                if (child.hasMutableChildren(attr)) {
                    return true;
                }
                if (child.localName === 'slot') {
                    return true;
                }
            }
        }
        return false;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/pointerevents2/#dom-element-haspointercapture
     */
    hasPointerCapture(pointerId) {
        throw new unexpected_call_error_1.default('Not supported "hasPointerCapture" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://html.spec.whatwg.org/multipage/popover.html#dom-hidepopover
     */
    hidePopover() {
        throw new unexpected_call_error_1.default('Not supported "hidePopover" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#dom-element-insertadjacentelement
     */
    insertAdjacentElement(where, 
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    element) {
        // TODO:
        throw new unexpected_call_error_1.default('Does not implement "insertAdjacentElement" method yet');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://w3c.github.io/DOM-Parsing/#widl-Element-insertAdjacentHTML-void-DOMString-position-DOMString-text
     */
    insertAdjacentHTML(position, text) {
        // TODO:
        throw new unexpected_call_error_1.default('Does not implement "insertAdjacentHTML" method yet');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#dom-element-insertadjacenttext
     */
    insertAdjacentText(where, data) {
        // TODO:
        throw new unexpected_call_error_1.default('Does not implement "insertAdjacentText" method yet');
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLElement`
     */
    isDescendantByUUIDList(uuidList) {
        let el = this.parentElement;
        if (el === null) {
            return false;
        }
        do {
            if (uuidList.includes(el.uuid)) {
                return true;
            }
            el = el.parentElement;
        } while (el !== null && el.is(el.ELEMENT_NODE));
        return false;
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLElement`
     */
    isEmpty() {
        var _a;
        for (const childNode of Array.from(this.childNodes)) {
            if (!(childNode.is(childNode.TEXT_NODE) && ((_a = childNode.textContent) === null || _a === void 0 ? void 0 : _a.trim()) === '')) {
                return false;
            }
        }
        return true;
    }
    /**
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-matches%E2%91%A0
     */
    matches(selector) {
        var _a;
        return ((0, selector_1.createSelector)(selector, this.ownerMLDocument.specs).match(
        // Prioritize the pretender
        ((_a = this.pretenderContext) === null || _a === void 0 ? void 0 : _a.type) === 'pretender' ? this.pretenderContext.as : this) !== false);
    }
    /**
     * Pretenders Initialization
     */
    pretending(pretenders) {
        var _a;
        const pretenderConfig = pretenders === null || pretenders === void 0 ? void 0 : pretenders.find(option => this.matches(option.selector));
        if (!pretenderConfig) {
            return;
        }
        let nodeName;
        let namespace = 'html';
        const attributes = [];
        let aria;
        if (typeof pretenderConfig.as === 'string') {
            nodeName = pretenderConfig.as;
        }
        else {
            nodeName = pretenderConfig.as.element;
            namespace = (_a = pretenderConfig.as.namespace) !== null && _a !== void 0 ? _a : namespace;
            if (pretenderConfig.as.inheritAttrs) {
                attributes.push(...this._astToken.attributes);
            }
            if (pretenderConfig.as.attrs) {
                attributes.push(...pretenderConfig.as.attrs.map(({ name, value }, i) => {
                    var _a;
                    const _value = value != null
                        ? typeof value === 'string'
                            ? value
                            : (_a = this.getAttribute(value.fromAttr)) !== null && _a !== void 0 ? _a : ''
                        : '';
                    return {
                        ...this._astToken,
                        uuid: `${this.uuid}_attr_${i}`,
                        type: 'html-attr',
                        nodeName: name,
                        spacesBeforeName: {
                            ...this._astToken,
                            raw: '',
                        },
                        name: {
                            ...this._astToken,
                            raw: name,
                        },
                        spacesBeforeEqual: {
                            ...this._astToken,
                            raw: '',
                        },
                        equal: {
                            ...this._astToken,
                            raw: '',
                        },
                        spacesAfterEqual: {
                            ...this._astToken,
                            raw: '',
                        },
                        startQuote: {
                            ...this._astToken,
                            raw: '',
                        },
                        value: {
                            ...this._astToken,
                            raw: _value,
                        },
                        endQuote: {
                            ...this._astToken,
                            raw: '',
                        },
                        isDuplicatable: true,
                        parentNode: null,
                        nextNode: null,
                        prevNode: null,
                        isFragment: false,
                        isGhost: false,
                    };
                }));
            }
            aria = pretenderConfig.as.aria;
        }
        const as = new MLElement({
            ...this._astToken,
            uuid: this.uuid + '_pretender',
            raw: `<${nodeName}>`,
            nodeName,
            namespace,
            elementType: 'html',
            attributes,
        }, this.ownerMLDocument);
        as.resetChildren(this.childNodes);
        as.pretenderContext = {
            type: 'origin',
            origin: this,
        };
        this.pretenderContext = {
            type: 'pretender',
            as,
            aria,
        };
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/pointerevents2/#dom-element-releasepointercapture
     */
    releasePointerCapture(pointerId) {
        throw new unexpected_call_error_1.default('Not supported "releasePointerCapture" method');
    }
    /**
     * @implements DOM API: `Element`
     */
    remove() {
        (0, child_node_methods_1.remove)(this);
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#dom-element-removeattribute
     */
    removeAttribute(qualifiedName) {
        throw new unexpected_call_error_1.default('Not supported "removeAttribute" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-removeattributens%E2%91%A0
     */
    removeAttributeNS(namespace, localName) {
        throw new unexpected_call_error_1.default('Not supported "removeAttributeNS" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/DOM-Level-2-Core/core.html#ID-D589198
     */
    removeAttributeNode(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    attr) {
        throw new unexpected_call_error_1.default('Not supported "removeAttributeNode" method');
    }
    /**
     * @implements DOM API: `Element`
     */
    replaceWith(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    ...nodes) {
        (0, child_node_methods_1.replaceWith)(this, ...nodes);
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://fullscreen.spec.whatwg.org/#dom-element-requestfullscreen
     */
    requestFullscreen(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    options) {
        throw new unexpected_call_error_1.default('Not supported "requestFullscreen" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://w3c.github.io/pointerlock/#dom-element-requestpointerlock
     */
    requestPointerLock() {
        throw new unexpected_call_error_1.default('Not supported "requestPointerLock" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/cssom-view-1/#dom-element-scroll
     */
    scroll(x, y) {
        throw new unexpected_call_error_1.default('Not supported "scroll" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/cssom-view-1/#dom-element-scrollby
     */
    scrollBy(x, y) {
        throw new unexpected_call_error_1.default('Not supported "scrollBy" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/cssom-view-1/#dom-element-scrollintoview
     */
    scrollIntoView(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    arg) {
        throw new unexpected_call_error_1.default('Not supported "scrollIntoView" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/cssom-view-1/#dom-element-scrollto
     */
    scrollTo(x, y) {
        throw new unexpected_call_error_1.default('Not supported "scrollTo" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-setattribute%E2%91%A0
     */
    setAttribute(qualifiedName, value) {
        throw new unexpected_call_error_1.default('Not supported "setAttribute" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-setattributens%E2%91%A0
     */
    setAttributeNS(namespace, qualifiedName, value) {
        throw new unexpected_call_error_1.default('Not supported "setAttributeNS" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#dom-element-setattributenode
     */
    setAttributeNode(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    attr) {
        throw new unexpected_call_error_1.default('Not supported "setAttributeNode" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#dom-element-setattributenodens
     */
    setAttributeNodeNS(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    attr) {
        throw new unexpected_call_error_1.default('Not supported "setAttributeNodeNS" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://www.w3.org/TR/pointerevents2/#idl-def-element-setpointercapture-pointerid
     */
    setPointerCapture(pointerId) {
        throw new unexpected_call_error_1.default('Not supported "setPointerCapture" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://html.spec.whatwg.org/multipage/popover.html#dom-showpopover
     */
    showPopover() {
        throw new unexpected_call_error_1.default('Not supported "showPopover" method');
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLElement`
     */
    toNormalizeString() {
        if (tslib_1.__classPrivateFieldGet(this, _MLElement_normalizedString, "f")) {
            return tslib_1.__classPrivateFieldGet(this, _MLElement_normalizedString, "f");
        }
        const children = this.getChildElementsAndTextNodeWithoutWhitespaces();
        const attrs = this.attributes.map(attr => attr.toNormalizeString());
        const attrString = attrs.length > 0 ? ' ' + attrs.join('') : '';
        const startTag = `<${this.nodeName}${attrString}>`;
        const childNodes = children.map(node => {
            if (node.is(node.ELEMENT_NODE)) {
                return node.toNormalizeString();
            }
            return node.originRaw;
        });
        const endTag = `</${this.nodeName}>`;
        const normalizedString = `${startTag}${childNodes.join('')}${endTag}`;
        tslib_1.__classPrivateFieldSet(this, _MLElement_normalizedString, normalizedString, "f");
        return normalizedString;
    }
    /**
     * @implements `@markuplint/ml-core` API: `MLElement`
     */
    toString() {
        return this.raw;
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://dom.spec.whatwg.org/#ref-for-dom-element-toggleattribute%E2%91%A0
     */
    toggleAttribute(qualifiedName, force) {
        throw new unexpected_call_error_1.default('Not supported "toggleAttribute" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @unsupported
     * @implements DOM API: `Element`
     * @see https://html.spec.whatwg.org/multipage/popover.html#dom-togglepopover
     */
    togglePopover(force) {
        throw new unexpected_call_error_1.default('Not supported "togglePopover" method');
    }
    /**
     * **IT THROWS AN ERROR WHEN CALLING THIS.**
     *
     * @deprecated
     * @unsupported
     * @implements DOM API: `Element`
     */
    webkitMatchesSelector(selectors) {
        throw new unexpected_call_error_1.default('Not supported "webkitMatchesSelector" method');
    }
}
exports.MLElement = MLElement;
_MLElement_attributes = new WeakMap(), _MLElement_fixedNodeName = new WeakMap(), _MLElement_getChildElementsAndTextNodeWithoutWhitespacesCache = new WeakMap(), _MLElement_localName = new WeakMap(), _MLElement_normalizedAttrs = new WeakMap(), _MLElement_normalizedString = new WeakMap(), _MLElement_tagOpenChar = new WeakMap();
