// @markuplint/rules/lib/required-attr/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ml_core_1 = require("@markuplint/ml-core");
const helpers_1 = require("../helpers");
exports.default = (0, ml_core_1.createRule)({
    defaultValue: [],
    async verify({ document, report, t }) {
        await document.walkOn('Element', el => {
            var _a;
            if (el.hasSpreadAttr) {
                return;
            }
            const customRequiredAttrs = typeof el.rule.value === 'string' ? [el.rule.value] : el.rule.value;
            const attrSpec = (0, ml_core_1.getAttrSpecs)(el, document.specs);
            const attributeSpecs = {};
            if (attrSpec && el.elementType === 'html') {
                for (const spec of attrSpec) {
                    if (spec.required != null || spec.requiredEither || spec.condition != null) {
                        attributeSpecs[spec.name] = {
                            ...spec,
                            values: [],
                        };
                    }
                }
            }
            for (const req of customRequiredAttrs) {
                let name;
                const values = [];
                if (typeof req === 'string') {
                    name = req;
                }
                else {
                    name = req.name;
                    if (Array.isArray(req.value)) {
                        values.push(...req.value);
                    }
                    else if (req.value) {
                        values.push(req.value);
                    }
                }
                attributeSpecs[name] = {
                    name,
                    values,
                    required: true,
                };
            }
            for (const spec of Object.values(attributeSpecs)) {
                const didntHave = !el.hasAttribute(spec.name);
                let invalid = false;
                if (spec.requiredEither) {
                    const candidate = [...spec.requiredEither, spec.name];
                    invalid = !candidate.some(attrName => el.hasAttribute(attrName));
                }
                else if (spec.required === true) {
                    invalid = (0, helpers_1.attrMatches)(el, spec.condition) && didntHave;
                }
                else if (spec.required != null && spec.required !== false) {
                    const selector = typeof spec.required === 'string' ? spec.required : spec.required.join(',');
                    invalid = el.matches(selector) && didntHave;
                }
                if (invalid) {
                    const message = t('{0} expects {1}', t('the "{0*}" {1}', el.localName, 'element'), t('the "{0*}" {1}', spec.name, 'attribute'));
                    report({
                        scope: el,
                        message,
                    });
                }
                const value = el.getAttribute(spec.name);
                if (spec.values.length > 0 && value) {
                    const matched = spec.values.some(pattern => (0, helpers_1.match)(value, pattern));
                    if (!matched) {
                        const expects = spec.values.length === 1 ? t(spec.values) : t('either {0}', t(spec.values));
                        const message = t('{0} expects {1}', t('the "{0*}" {1}', spec.name, 'attribute'), expects);
                        const attrToken = el.getAttributeToken(spec.name);
                        const valueToken = (_a = attrToken[0]) === null || _a === void 0 ? void 0 : _a.valueNode;
                        const token = valueToken !== null && valueToken !== void 0 ? valueToken : attrToken[0];
                        if (token) {
                            report({
                                scope: {
                                    rule: el.rule,
                                    raw: token.raw,
                                    startCol: token.startCol,
                                    startLine: token.startLine,
                                },
                                message,
                            });
                        }
                    }
                }
            }
        });
    },
});
