// @markuplint/rules/lib/no-use-event-handler-attr/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ml_core_1 = require("@markuplint/ml-core");
const helpers_1 = require("../helpers");
exports.default = (0, ml_core_1.createRule)({
    defaultSeverity: 'warning',
    defaultOptions: {},
    async verify({ document, report, t }) {
        await document.walkOn('Attr', attr => {
            if (attr.ownerElement.elementType !== 'html') {
                return;
            }
            const ignoreList = Array.isArray(attr.rule.options.ignore)
                ? attr.rule.options.ignore
                : attr.rule.options.ignore
                    ? [attr.rule.options.ignore]
                    : [];
            const name = attr.name;
            for (const ignore of ignoreList) {
                if ((0, helpers_1.match)(name, ignore)) {
                    return;
                }
            }
            if (/^on/i.test(name)) {
                report({
                    scope: attr,
                    raw: attr.raw,
                    line: attr.startLine,
                    col: attr.startCol,
                    message: t('{0} is disallowed', t('the "{0*}" {1}', name, 'attribute')),
                });
            }
        });
    },
});
