// @markuplint/rules/lib/no-default-value/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ml_core_1 = require("@markuplint/ml-core");
const helpers_1 = require("../helpers");
exports.default = (0, ml_core_1.createRule)({
    defaultSeverity: 'warning',
    async verify({ document, report, t }) {
        await document.walkOn('Attr', attr => {
            var _a, _b, _c;
            const attrSpec = (0, ml_core_1.getAttrSpecs)(attr.ownerElement, document.specs);
            if (!attrSpec) {
                return;
            }
            const spec = attrSpec.find(s => s.name === attr.name);
            if (!spec) {
                return;
            }
            if (!spec.defaultValue) {
                return;
            }
            const normalizedValue = (0, helpers_1.toNormalizedValue)(attr.value, spec);
            const defaultValue = (0, helpers_1.toNormalizedValue)(spec.defaultValue, spec);
            if (defaultValue === normalizedValue) {
                report({
                    scope: attr,
                    line: (_a = attr.valueNode) === null || _a === void 0 ? void 0 : _a.startLine,
                    col: (_b = attr.valueNode) === null || _b === void 0 ? void 0 : _b.startCol,
                    raw: (_c = attr.valueNode) === null || _c === void 0 ? void 0 : _c.raw,
                    message: t('It is {0}', t('the {0}', 'default value')),
                });
            }
        });
    },
});
