// @markuplint/rules/lib/attr-duplication/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ml_core_1 = require("@markuplint/ml-core");
exports.default = (0, ml_core_1.createRule)({
    async verify({ document, report, t }) {
        const message = t('{0} is {1:c}', t('the {0}', 'attribute name'), 'duplicated');
        await document.walkOn('Element', node => {
            var _a;
            const attrNameStack = [];
            for (const attr of node.getAttributeTokens()) {
                if (attr.isDuplicatable) {
                    continue;
                }
                const attrName = attr.name;
                const name = node.elementType === 'html' ? attrName.toLowerCase() : attrName;
                if (attrNameStack.includes(name)) {
                    const scope = (_a = attr.nameNode) !== null && _a !== void 0 ? _a : attr;
                    report({
                        scope: {
                            rule: node.rule,
                            startLine: scope.startLine,
                            startCol: scope.startCol,
                            raw: scope.raw,
                        },
                        message,
                    });
                }
                else {
                    attrNameStack.push(name);
                }
            }
        });
    },
});
