// @markuplint/ml-core/lib/ml-dom/helper/debug.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodeListToDebugMaps = void 0;
function nodeListToDebugMaps(
// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
nodeList, withAttr = false) {
    return nodeList
        .map(n => {
        const r = [];
        if (n.is(n.ELEMENT_NODE) && n.isOmitted) {
            r.push(`[N/A]>[N/A](N/A)${n.nodeName}: ${visibleWhiteSpace(n.raw)}`);
            return r;
        }
        r.push(tokenDebug(n));
        if (n.is(n.ELEMENT_NODE)) {
            r.push(`  namespaceURI: ${!!n.namespaceURI}`);
            r.push(`  elementType: ${n.elementType}`);
            r.push(`  isInFragmentDocument: ${n.isInFragmentDocument()}`);
            r.push(`  isForeignElement: ${!!n.isForeignElement}`);
            if (withAttr) {
                r.push(...attributesToDebugMaps(n.attributes)
                    .flat()
                    .map(l => `  ${l}`));
            }
        }
        return r;
    })
        .flat();
}
exports.nodeListToDebugMaps = nodeListToDebugMaps;
function attributesToDebugMaps(
// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
attributes) {
    return attributes
        .map(n => {
        const r = [
            tokenDebug({
                name: n.name,
                startOffset: n.startOffset,
                endOffset: n.endOffset,
                startLine: n.startLine,
                endLine: n.endLine,
                startCol: n.startCol,
                endCol: n.endCol,
                raw: n.raw,
            }),
        ];
        if (n.spacesBeforeName) {
            r.push(`  ${tokenDebug(n.spacesBeforeName, 'bN')}`);
        }
        if (n.nameNode) {
            r.push(`  ${tokenDebug(n.nameNode, 'name')}`);
        }
        if (n.spacesBeforeEqual) {
            r.push(`  ${tokenDebug(n.spacesBeforeEqual, 'bE')}`);
        }
        if (n.equal) {
            r.push(`  ${tokenDebug(n.equal, 'equal')}`);
        }
        if (n.spacesAfterEqual) {
            r.push(`  ${tokenDebug(n.spacesAfterEqual, 'aE')}`);
        }
        if (n.startQuote) {
            r.push(`  ${tokenDebug(n.startQuote, 'sQ')}`);
        }
        if (n.valueNode) {
            r.push(`  ${tokenDebug(n.valueNode, 'value')}`);
        }
        if (n.endQuote) {
            r.push(`  ${tokenDebug(n.endQuote, 'eQ')}`);
        }
        if (n.spacesBeforeName) {
            r.push(`  isDirective: ${!!n.isDirective}`);
            r.push(`  isDynamicValue: ${!!n.isDynamicValue}`);
        }
        if (n.candidate) {
            r.push(`  candidate: ${visibleWhiteSpace(n.candidate)}`);
        }
        return r;
    })
        .flat();
}
function tokenDebug(n, type = '') {
    var _a, _b, _c, _d;
    return `[${n.startLine}:${n.startCol}]>[${n.endLine}:${n.endCol}](${n.startOffset},${n.endOffset})${(_d = (_c = (_b = (_a = n.nodeName) !== null && _a !== void 0 ? _a : n.potentialName) !== null && _b !== void 0 ? _b : n.name) !== null && _c !== void 0 ? _c : n.type) !== null && _d !== void 0 ? _d : type}: ${visibleWhiteSpace(n.raw)}`;
}
function visibleWhiteSpace(chars) {
    return chars.replace(/\n/g, '⏎').replace(/\t/g, '→').replace(/\s/g, '␣');
}
