// @markuplint/rules/lib/permitted-contents/content-model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.contentModel = void 0;
const ml_spec_1 = require("@markuplint/ml-spec");
const start_1 = require("./start");
function contentModel(
// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
el, rules, options) {
    const model = createModel(el, rules);
    if (model == null) {
        return [
            {
                type: 'MATCHED',
                scope: el,
                query: '*',
                hint: {},
            },
        ];
    }
    const result = (0, start_1.start)(model, el, el.ownerMLDocument.specs, options);
    return result;
}
exports.contentModel = contentModel;
function createModel(
// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
el, rules) {
    const specs = cachedSpecs(el.ownerMLDocument.specs, rules);
    const model = (0, ml_spec_1.getContentModel)(el, specs.specs);
    return model;
}
const caches = new Map();
function cachedSpecs(specs, rules) {
    if (rules.length === 0) {
        return specs;
    }
    const key = JSON.stringify(rules);
    const cached = caches.get(key);
    if (cached) {
        return cached;
    }
    const merged = {
        ...specs,
        specs: [
            ...specs.specs,
            ...rules.map(tag => ({
                name: tag.tag,
                contentModel: tag,
            })),
        ],
    };
    caches.set(key, merged);
    return merged;
}
