// @markuplint/rules/lib/permitted-contents/debug.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cyan = exports.bgMagenta = exports.blue = exports.bgBlue = exports.bgRed = exports.green = exports.bgGreen = exports.cmLog = void 0;
const debug_1 = require("../debug");
exports.cmLog = debug_1.log.extend('content-model');
var ansi_colors_1 = require("ansi-colors");
Object.defineProperty(exports, "bgGreen", { enumerable: true, get: function () { return ansi_colors_1.bgGreen; } });
Object.defineProperty(exports, "green", { enumerable: true, get: function () { return ansi_colors_1.green; } });
Object.defineProperty(exports, "bgRed", { enumerable: true, get: function () { return ansi_colors_1.bgRed; } });
Object.defineProperty(exports, "bgBlue", { enumerable: true, get: function () { return ansi_colors_1.bgBlue; } });
Object.defineProperty(exports, "blue", { enumerable: true, get: function () { return ansi_colors_1.blue; } });
Object.defineProperty(exports, "bgMagenta", { enumerable: true, get: function () { return ansi_colors_1.bgMagenta; } });
Object.defineProperty(exports, "cyan", { enumerable: true, get: function () { return ansi_colors_1.cyan; } });
