// @markuplint/rules/lib/permitted-contents/transparent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transparent = void 0;
const debug_1 = require("./debug");
const transparentLog = debug_1.cmLog.extend('transparent');
function transparent(
// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
nodes) {
    var _a, _b;
    if (nodes.length === 0 || ((_b = (_a = nodes[0]) === null || _a === void 0 ? void 0 : _a.parentElement) === null || _b === void 0 ? void 0 : _b.parentElement)) {
        transparentLog('Skipped');
        return {
            type: nodes.length === 0 ? 'MATCHED_ZERO' : 'MATCHED',
            matched: nodes.slice(),
            unmatched: [],
            zeroMatch: nodes.length === 0,
            query: 'transparent',
            hint: {},
        };
    }
    transparentLog('Transparent model element is component root');
    return {
        type: 'MATCHED',
        matched: nodes.slice(),
        unmatched: [],
        zeroMatch: false,
        query: 'transparent',
        hint: {},
    };
}
exports.transparent = transparent;
